pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ControllerLeaderLocatorIntegrationTest.java

```diff
@@ -0,0 +1,164 @@
+    TestUtils.waitForCondition(aVoid -> {
+      String tableName;
+      try {
+        tableName = generateTableNameUsingSecondController(secondControllerConfig, testTableName1);
+      } catch (Exception e) {
+        return false;
+      }
+      testTableName3.set(tableName);
+      return tableName != null;
+    }, TIMEOUT_IN_MS, "Failed to find the second table");
+
+    // Mock the behavior that 40 seconds have passed.
+    controllerLeaderLocator.setLastCacheInvalidateMillis(System.currentTimeMillis() - 40_000L);
+    controllerLeaderLocator.invalidateCachedControllerLeader();
+
+    // The second controller should be the lead controller for test table 3, which isn't the helix leader.
+    Pair<String, Integer> thirdPair = controllerLeaderLocator.getControllerLeader(testTableName3.get());
+    Assert.assertNotNull(thirdPair);
+    Assert.assertEquals(thirdPair.getFirst(), ControllerTest.LOCAL_HOST);
+    Assert.assertEquals((int) thirdPair.getSecond(), (ControllerTest.DEFAULT_CONTROLLER_PORT + 1));
+    Assert.assertNotEquals(pair.getSecond(), thirdPair.getSecond());
+
+    secondControllerStarter.stop();
+  }
+
+  /**
+   * Generates a table name which uses the second controller as its lead controller.
+   */
+  private String generateTableNameUsingSecondController(ControllerConf secondControllerConfig, String testTableName) {
+    String secondParticipantId = LeadControllerUtils
+        .generateParticipantInstanceId(secondControllerConfig.getControllerHost(),
+            Integer.parseInt(secondControllerConfig.getControllerPort()));
+    Set<Integer> partitionIdsForSecondController = new HashSet<>();
+
+    // Find out all the partitionIds that the second controller is assigned to.
+    int partitionId = LeadControllerUtils.getPartitionIdForTable(testTableName);
+    String partitionNameForTable1 = LeadControllerUtils.generatePartitionName(partitionId);
+    ExternalView leadControllerResourceExternalView =
+        _helixAdmin.getResourceExternalView(_helixManager.getClusterName(), Helix.LEAD_CONTROLLER_RESOURCE_NAME);
+    Set<String> partitionNames = leadControllerResourceExternalView.getPartitionSet();
+    for (String partitionName : partitionNames) {
+      Map<String, String> partitionStateMap = leadControllerResourceExternalView.getStateMap(partitionName);
+      // Get master host from partition map. Return null if no master found.
+      for (Map.Entry<String, String> entry : partitionStateMap.entrySet()) {
+        if (!partitionNameForTable1.equals(partitionName) && MasterSlaveSMD.States.MASTER.name()
+            .equals(entry.getValue())) {
+          if (secondParticipantId.equals(entry.getKey())) {
+            partitionIdsForSecondController.add(LeadControllerUtils.extractPartitionId(partitionName));
+          }
+        }
+      }
+    }
+
+    // Try 100 times to find out a table which is using the second controller as lead controller.
+    for (int i = 0; i < 100; i++) {
```

Comment:
pre-generate table names since the number of partitions are fixed.

Reply:
Nope. Because we don't know whether the pre-generated table name would get the same controller leader even though two tables belong to two different partitions

Reply:
Refactored the test to generate a partition-tableName map.